src/dfcx_scrapi/core/test_cases.py

```diff
@@ -430,3 +432,149 @@ def calculate_coverage(self, coverage_type: int, agent_id: str = None):
+        # TODO: Should throw error, but returning this probably will anyway
+        return "Invalid"
+
+    def get_test_case_results_df(self, agent_id=None, retest_all=False):
+        """Gets the test case results for this agent,
+        and generates a dataframe with their details.
+        Any tests without a result will be run in a batch.
+
+        Args:
+          agent_id: required only if not set when initializing this class
+          retest_all: if true, all test cases are re-run,
+            regardless of whether or not they had a result
+
+        Returns:
+          DataFrame of test case results for this agent, with columns:
+            display_name, id, short_id (excluding agent ID),
+            tags (comma-separated string), creation_time,
+            start_flow, start_page, passed, test_time
+        """
+        if agent_id:
+            self.agent_id = agent_id
+
+        dfcx_flows = Flows(creds=self.creds, agent_id=self.agent_id)
+        dfcx_pages = Pages(creds=self.creds)
+        flows_map = dfcx_flows.get_flows_map(agent_id=self.agent_id)
+        pages_map = {}
+        for flow_id in flows_map.keys():
+            pages_map[flow_id] = dfcx_pages.get_pages_map(flow_id=flow_id)
+
+        test_case_results = self.list_test_cases(self.agent_id)
+        retest = []
+        retest_names = []
+
+        display_names = []
+        ids = []
+        short_ids = []
+        tags = []
+        creation_times = []
+        flows = []
+        pages = []
+        test_results = []
+        test_times = []
+        passed = []
+
+        for response in test_case_results:
+            # Collect untested cases to be retested
+            # (or all if retest_all is True)
+            if (
+                retest_all
+                or str(response.last_test_result.test_result)
+                == "TestResult.TEST_RESULT_UNSPECIFIED"
+            ):
+                retest.append(response.name)
+                retest_names.append(response.display_name)
+                # Collect additional information for dataframe
+            display_names.append(response.display_name)
+            ids.append(response.name)
+            short_ids.append(response.name.split("/")[-1])
+            tags.append(",".join(response.tags))
+            creation_times.append(response.creation_time)
+            flows.append(self._convert_flow(response.test_config.flow,
+                                            flows_map)
+            )
+            pages.append(
+                self._convert_page(response.test_config.page,
+                                   response.test_config.flow,
+                                   pages_map)
+            )
+            test_results.append(str(response.last_test_result.test_result))
+            test_times.append(response.last_test_result.test_time)
+            passed.append(
+                str(response.last_test_result.test_result)=="TestResult.PASSED"
+            )
+
+        # Create dataframe
```

Comment:
@SeanScripts Same here. 

Break this out into a simple method to make the dataframe and cut it down to a 1-liner in your main function here.

```py
test_case_df = self.make_dataframe(args)
```

You want `get_test_case_results_df` to look really clean and mostly have lines that are the functional components of what is happening. Makes it easier to debug later.

Reply:
By making the above section a new function, I was able to make the dataframe by just concatenating the list of rows